value.go

```diff
@@ -196,3 +208,39 @@ func (BinaryValueCodec) UnmarshalValueKeys(b []byte) ([][]byte, error) {
 	}
 	return vk, nil
 }
+
+func (bjc BinaryWithJsonFallbackCodec) MarshalValue(v Value) ([]byte, error) {
+	return bjc.BinaryValueCodec.MarshalValue(v)
```

Comment:
That would break the current binary format. No?

Reply:
correct - it would be a different one. i just wonder if we believe the current setup is safe, or if we might be able to have new values written as binary that should be recovered as binary but are instead successfully trial-decrypted as json.

Reply:
We could introduce a new format. But we still need to be be able to deal with gracefully decoding the two current ones.

What if the check for whether to decode as binary or not reads a variant first and checks that the remaining byte length matches the value of variant. If it does it's binary, if not JSON. Can you think of any edge cases that'd leave behind?

Reply:
does the current binary format start with a varint?

```
	n, _ := binary.Uvarint([]byte{'{'})
	fmt.Printf("val is %d", n)
```
```
val is 123
```

Reply:
It [does](https://github.com/filecoin-project/go-indexer-core/blob/main/value.go#L109). 
So in the case above we know the value is JSON if its length is less than 122. We can also do a further check for JSON object closed at last byte.

WDYT?

Reply:
it would be good if when using this, someone can't write a malicious binary value that is subsequently interpreted as json

Reply:
So the problem is when there is a value that looks like `{ ... }` where the `...` is 121 bytes of data. It is either a json value that looks like binary or a binary value that looks like json. All other cases are unambiguous.

If we do not have any such existing cases in our value store, before introducing the new codec, then this can safely be interpreted as binary. If any of these already exist, then they will be ambiguous to the new code, unless there is some other way to tell them apart.

If the chance that there were any json values matching this is nearly nonexistent, then interpret these as binary and deal with corrupted decoding if ever it should occur.

Reply:
I think that's right - if we can have the edge case be decoded as the new format we should be safe.

Reply:
That's what the code does right now. no? Where we attempt to decode as JSON first and if it fails fall back on binary

Reply:
In terms of malicious/accidental behaviour, there is a chance that a value can be both a valid binary format and a valid JSON format.

In this case there is no way to differentiate which was which unless we introduce some sentinel value, i.e. yet another format. maybe that's unlikely enough to ignore?

Reply:
I think the current one in the ambiguous case interprets it as the old format (JSON)

Reply:
but i think your proposal of first checking that the initial uvarint matches length and then doing binary if so means the ambiguous case will be interpreted as the new format (Binary), so that should make this safe

Reply:
I see